### PR TITLE
Switch from flutter_blue to flutter_reactive_ble

### PR DESCRIPTION
**Pull request by MouradLachhab**
Switch was made to flutter reactive ble framework allowing to set the BLE priority which drops the packet loss considerably. However, this framework does not provide methods to disconnect a connected device. Error handling was also changed, I'm not sure if this is the best way to do it.

### CODE REVIEW

File: mobile/lib/src/application/device/device_selector_cubit.dart [hunk 1]
```diff
@@ -30,13 +30,26 @@ class DeviceSelectorCubit extends Cubit<DeviceState> {
   Future<void> connect(AcquisitionDevice device) async {
     emit(DeviceConnectionInProgress());
 
+<<<<<<< HEAD
```
Comment (abelfodil) [inline mobile/lib/src/application/device/device_selector_cubit.dart:33]: conflict

File: mobile/lib/src/domain/eeg_data/eeg_data.dart [hunk 2]
```diff
@@ -2,8 +2,12 @@
 import 'package:polydodo/src/domain/entity.dart';
 
 class EEGData extends Entity {
+<<<<<<< HEAD
```
Comment (abelfodil) [inline mobile/lib/src/domain/eeg_data/eeg_data.dart:5]: conflict

**Review by abelfodil — CHANGES_REQUESTED** [review]
Pls fix conflicts

File: mobile/lib/src/application/device/device_selector_cubit.dart (hunk 1 above)
Comment (MouradLachhab) [reply mobile/lib/src/application/device/device_selector_cubit.dart:33]: I noticed, I'm fixing them right now. It's weird they weren't showing up in my vscode, clean open everything.

File: mobile/lib/src/domain/acquisition_device/i_acquisition_device_repository.dart [hunk 3]
```diff
@@ -3,7 +3,8 @@ import 'acquisition_device.dart';
 abstract class IAcquisitionDeviceRepository {
   void initializeRepository();
 
-  Future<void> connect(AcquisitionDevice device);
+  Future<void> connect(
+      AcquisitionDevice device, Function(bool, Exception) callback);
```
Comment (abelfodil) [inline mobile/lib/src/domain/acquisition_device/i_acquisition_device_repository.dart:7]:
```suggestion
  void connect(AcquisitionDevice device, Function(bool, Exception) callback);
```

File: mobile/lib/src/infrastructure/bluetooth_repository.dart [hunk 4]
```diff
@@ -1,129 +1,112 @@
...
     }
 
     streamController.add(_acquisitionDevicePersistency);
   }
 
-  Future<void> connect(AcquisitionDevice device) async {
-    _selectedDevice =
-        _bluetoothDevices[_acquisitionDevicePersistency.indexOf(device)];
+  Future<void> connect(
+      AcquisitionDevice device, Function(bool, Exception) callback) async {
```
Comment (abelfodil) [inline mobile/lib/src/infrastructure/bluetooth_repository.dart:53]:
```suggestion
  void connect(AcquisitionDevice device, Function(bool, Exception) callback) {
```

**Review by abelfodil — CHANGES_REQUESTED** [review]

File: mobile/lib/src/infrastructure/bluetooth_repository.dart [hunk 5]
```diff
@@ -1,129 +1,112 @@
...
-        .characteristics;
-    for (BluetoothCharacteristic characteristic in characteristics) {
-      if (characteristic.uuid.toString().contains(BLE_RECEIVE)) {
-        _receiveCharacteristic = characteristic;
-      } else if (characteristic.uuid.toString().contains(BLE_SEND)) {
-        _sendCharacteristic = characteristic;
-      }
-    }
+  Future<void> setupCharacteristics() async {
+    _sendCharacteristic = QualifiedCharacteristic(
```
Comment (abelfodil) [inline mobile/lib/src/infrastructure/bluetooth_repository.dart:82]: vraiment plus clean :+1:

File: mobile/lib/src/infrastructure/eeg_data_repository.dart [hunk 6]
```diff
@@ -30,13 +27,14 @@ class EEGDataRepository implements IEEGDataRepository {
...
   Future<void> stopRecordingFromStream() async {
     // todo: move save future to another file
+
     final directory = await getExternalStorageDirectory();
     final pathOfTheFileToWrite =
-        directory.path + '/' + recordingData.fileName + ".txt";
+        directory.path + '/' + _recordingData.fileName + ".txt";
     File file = File(pathOfTheFileToWrite);
-    List<List> fileContent = new List<List>();
+    List<List> fileContent = List<List>();
```
Comment (abelfodil) [inline mobile/lib/src/infrastructure/eeg_data_repository.dart:35]: nit: on devrait prendre l'habitude de d'utiliser `var`
nit++ (préférence personnelle): `[]` au lieu de `List<List>()`

File: mobile/pubspec.yaml [hunk 7]
```diff
@@ -42,6 +41,7 @@ dependencies:
   binary: ^2.0.0
   csv: ^4.0.3
   path_provider: ^1.6.16
+  flutter_reactive_ble : ^2.5.2
```
Comment (abelfodil) [inline mobile/pubspec.yaml:44]: nit: sort this list

**Comment by abelfodil** [comment]
lgtm sinon, tu peux ignorer les nit si tu veux, à toi de choisir.

File: mobile/lib/src/infrastructure/bluetooth_repository.dart [hunk 8]
```diff
@@ -1,129 +1,112 @@
...
+
+    stream = flutterReactiveBle
+        .connectToDevice(id: _selectedDevice.id.toString())
+        .timeout(Duration(seconds: 10), onTimeout: (sink) {
+      sink.close();
+      callback(false, Exception("Connection Timed out"));
+    }).listen((event) {
+      if (event.connectionState == DeviceConnectionState.connected) {
+        setupCharacteristics();
+        stream.cancel();
```
Comment (abelfodil) [inline mobile/lib/src/infrastructure/bluetooth_repository.dart:68]: https://github.com/PhilipsHue/flutter_reactive_ble/issues/27#issuecomment-590329365

**Review by abelfodil — CHANGES_REQUESTED** [review]
This code seems to disconnect the device, so we should preserve the stream value and use it to disconnect by cancelling the stream elsewhere.

**Review by abelfodil — APPROVED** [review]

File: mobile/lib/src/infrastructure/bluetooth_repository.dart [hunk 9]
```diff
@@ -1,129 +1,112 @@
...
+      } else if (event.connectionState == DeviceConnectionState.disconnected) {
+        disconnect();
+        callback(false, Exception("Failed to connect to device"));
+      }
+    });
   }
 
-  Future<void> disconnect() async {
+  void disconnect() async {
     if (_selectedDevice != null) {
```
Comment (abelfodil) [inline mobile/lib/src/infrastructure/bluetooth_repository.dart:75]: https://dart.dev/codelabs/dart-cheatsheet#conditional-property-access

**Pull request merged**